store/tikv/batch_coprocessor.go

```diff
@@ -0,0 +1,422 @@
+	}
+	b.wg.Wait()
+	return nil
+}
+
+func (b *batchCopIterator) handleTask(ctx context.Context, bo *Backoffer, task *batchCopTask) {
+	logutil.BgLogger().Debug("handle batch task")
+	tasks := []*batchCopTask{task}
+	idx := 0
+	for idx < len(tasks) {
+		ret, err := b.handleTaskOnce(ctx, bo, task)
+		if err != nil {
+			resp := &batchCopResponse{err: errors.Trace(err)}
+			b.sendToRespCh(resp)
+		} else {
+			tasks = append(tasks, ret...)
+		}
+		idx++
+	}
+	b.wg.Done()
+}
+
+// Merge all ranges and request again.
+func (b *batchCopIterator) retryBatchCopTask(ctx context.Context, bo *Backoffer, batchTask *batchCopTask) ([]*batchCopTask, error) {
+	ranges := &copRanges{}
+	for _, taskCtx := range batchTask.copTasks {
+		taskCtx.task.ranges.do(func(ran *kv.KeyRange) {
+			ranges.mid = append(ranges.mid, *ran)
+		})
+	}
+	return buildBatchCopTasks(bo, b.RegionCache, ranges, b.req)
+}
+
+func (b *batchCopIterator) handleTaskOnce(ctx context.Context, bo *Backoffer, task *batchCopTask) ([]*batchCopTask, error) {
+	logutil.BgLogger().Debug("handle batch task once")
+	sender := NewRegionBatchRequestSender(b.store.regionCache, b.store.client)
+	var regionInfos []*coprocessor.RegionInfo
+	for _, task := range task.copTasks {
+		regionInfos = append(regionInfos, &coprocessor.RegionInfo{
+			RegionId: task.task.region.id,
+			RegionEpoch: &metapb.RegionEpoch{
+				ConfVer: task.task.region.confVer,
+				Version: task.task.region.ver,
+			},
+			Ranges: task.task.ranges.toPBRanges(),
+		})
+	}
+
+	copReq := coprocessor.BatchRequest{
+		Tp:        b.req.Tp,
+		StartTs:   b.req.StartTs,
+		Data:      b.req.Data,
+		SchemaVer: b.req.SchemaVar,
+		Regions:   regionInfos,
+	}
+
+	req := tikvrpc.NewReplicaReadRequest(task.cmdType, &copReq, b.req.ReplicaRead, b.replicaReadSeed, kvrpcpb.Context{
+		IsolationLevel: pbIsolationLevel(b.req.IsolationLevel),
+		Priority:       kvPriorityToCommandPri(b.req.Priority),
+		NotFillCache:   b.req.NotFillCache,
+		HandleTime:     true,
+		ScanDetail:     true,
+	})
+	req.StoreTp = kv.TiFlash
+
+	logutil.BgLogger().Debug("send batch request to ", zap.String("req info", req.String()), zap.Int("cop task len", len(task.copTasks)))
+	resp, retry, err := sender.sendReqToAddr(bo, task.copTasks, req, ReadTimeoutMedium)
+	// If there are store errors, we should retry for all regions.
+	if retry {
+		return b.retryBatchCopTask(ctx, bo, task)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return b.handleStreamedBatchCopResponse(ctx, bo, resp.Resp.(*tikvrpc.BatchCopStreamResponse), task)
+}
+
+func (b *batchCopIterator) handleStreamedBatchCopResponse(ctx context.Context, bo *Backoffer, response *tikvrpc.BatchCopStreamResponse, task *batchCopTask) (totalRetTask []*batchCopTask, err error) {
+	defer response.Close()
+	resp := response.BatchResponse
+	if resp == nil {
+		// streaming request returns io.EOF, so the first Response is nil.
+		return nil, nil
+	}
+	for {
+		remainedTasks, err := b.handleBatchCopResponse(bo, resp, task)
```

Comment:
`remainedTasks` is always empty? I see the function `handleBatchCopResponse` return an empty slice.

Reply:
yes, i will fix it